File: c++/dynamicProgramming/1-D.cpp
#include <map>
using namespace std;

class Solution
{
public:
    int climbStairs(int n)
    {

        /* bottom up apprach O(n) time and O(1) space */
        if (n <= 2)
        {
            return n;
        }
        int a[2] = {1, 2};

        for (int i = 3; i < n + 1; i++)
        {
            int tmp(a[1]);
            a[1] = a[0] + a[1];
            a[0] = tmp;
        }
        return a[1];

        /* Memorizatoin approach. O(n) time and space */
        // if (n <= 2){
        //     return n;
        // }
        // if (arr.find(n) != arr.end()){
        //     return arr[n];
        // }
        // else {
        //     arr[n] = climbStairs(n - 1) + climbStairs(n - 2);
        // }
        // return arr[n];

        /* Brute force approach. O(2^n) time and space O(n) */
        // if (n <= 2){
        //     return n;
        // }

        // return climbStairs(n - 1) + climbStairs(n - 2);
    }

private:
    map<int, int> arr;
};

File: Python/DynamicProgramming/1-D.py
""" Climbing stairs
"""


class Solution:
    def __init__(self):
        self.cache = {}

    def climbStairs(self, n: int) -> int:
        """bottom up dynamic programming O(n) time and O(1) space"""
        if n <= 2:
            return n
        
        arr = [1, 2]
        for i in range(3, n + 1):
            temp = arr[1]
            arr[1] = arr[0] + arr[1]
            arr[0] = temp
        
        return arr[1]
        


        """top down dynamic programming using memorization O(n) time and O(n) space"""
        # if n <= 2:
        #     return n
        # elif n in self.cache:
        #     return self.cache[n]
        
        # else:
        #     self.cache[n] = self.climbStairs(n - 1) + self.climbStairs(n - 2)
        
        # return self.cache[n]

        """Brute Force O(2^n) with O(n) space"""
        # if n <=2:
        #     return n
        # return self.climbStairs(n - 1) + self.climbStairs(n - 2)
        